src/datasets/arrow_dataset.py

```diff
@@ -3331,6 +3335,146 @@ def to_parquet(
+        else:
+            dataset_name = identifier[0]
+            organization = None
+
+        api = HfApi()
+
+        try:
+            api.create_repo(
+                dataset_name,
+                token,
+                repo_type="dataset",
+                organization=organization,
+                private=private,
+            )
+        except HTTPError as err:
+            if err.response.status_code == 409:
+                if private is not None:
+                    logger.warning("The repository already exists: the `private` keyword argument will be ignored.")
+            else:
+                raise
+
+        if self._indices is not None:
+            dataset_nbytes = self.data.nbytes * len(self._indices) / len(self.data)
+        else:
+            dataset_nbytes = self.data.nbytes
+
+        num_shards = int(dataset_nbytes / shard_size) + 1
+        shards = (self.shard(num_shards=num_shards, index=i, contiguous=True) for i in range(num_shards))
+
+        files = api.list_repo_files(repo_id, repo_type="dataset", revision=branch, token=token)
+        files = [file for file in files if file.startswith("data/")]
+
+        def path_in_repo(_index):
+            return f"data/{split}-{_index:05d}-of-{num_shards:05d}.parquet"
+
+        # Only delete file shards that don't currently exist. Others will be overwritten if the content is different
+        # or will be left intact is the content is identical.
+        def should_delete_file(file_name):
+            file_to_overwrite = file_name in [path_in_repo(i) for i in range(num_shards)]
+            file_from_same_split = file_name.startswith(f"data/{split}-")
+
+            return file_from_same_split and not file_to_overwrite
+
+        file_shards_to_delete = [file for file in files if should_delete_file(file)]
+
+        def delete_file(file):
+            api.delete_file(file, repo_id=repo_id, token=token, repo_type="dataset", revision=branch)
+
+        if len(file_shards_to_delete):
+            # Making this available just for testing purposes: the moon-staging endpoint is having a super hard time
+            # digesting concurrent requests. Locally, I get error 500 returned 90% of the time when threading.
+            # would like to confirm this is only related to moon-staging before removing the ability to toggle threading
+            # on/off.
+            if threading:
+                thread_map(
+                    delete_file,
+                    file_shards_to_delete,
+                    desc="Deleting unused files from dataset repository",
+                    total=len(file_shards_to_delete),
+                )
+            else:
+                for file in utils.tqdm(
+                    file_shards_to_delete,
+                    desc="Deleting unused files from dataset repository",
+                    total=len(file_shards_to_delete),
+                    disable=bool(logging.get_verbosity() == logging.NOTSET),
```

Comment:
A tiny nit:

```suggestion
                    disable=bool(logging.get_verbosity() == logging.NOTSET) or not utils.is_progress_bar_enabled(),
```

Feel free to ignore it because we are not correctly handling `disable` in `thread_map` in one more place in the codebase, so this will need a separate PR.